### ISSUE TITLE
continous integration (CI) on gitlab

### ISSUE THREAD

**Issue by jpata** [vhbb/cmssw#574]
CERN gitlab supports native CI on CERN OpenStack: https://gitlab.cern.ch/ttH/CommonClassifier/commit/7ef6fb4beea5bc5cf1c78a9cac97e30d306df6ce/builds

CI means that a machine will automatically test a PR once it's made, before it's merged. I would propose to move the vhbb repo to gitlab, once we figure out the new model.

**Comment by arizzi** [comment]
do you know how central cms-sw does it on github.com ?  anyhow I'm ok with trying the gitlab option if someone is going to maintain the autotesting scripts

**Comment by jpata** [comment]
I believe cms-sw has it's own tests system at CERN via the integration builds, I don't really know the details there. The way to set up this testing on openstack seems relatively straightforward, I'm doing it for my own analysis codes now and I'll report how it went.

**Comment by arizzi** [comment]
great

**Comment by ferencek** [comment]
Naively, I would expect that using this CI feature will only be possible if the official repo is also migrated to Gitlab.

**Comment by jpata** [comment]
why? in git there is no concept of "official repo".

**Comment by ferencek** [comment]
In theory. In practice you are still making PRs to just one, main repo.

**Comment by jpata** [comment]
well, if you want your PR-s to be tested automatically, then indeed you need to PR against the repo where CI is set up. I guess by official repo you mean `github.com/vhbb/cmssw` that the vhbb-folks use for their activities?

**Comment by ferencek** [comment]
Yes, here the "official" repo would actually be https://github.com/vhbb/cmssw since PRs are made against it. So in principle it could be migrated to Gitlab and then later somebody would have to take care of keeping the code in sync with the upstream developments in https://github.com/cms-sw/cmssw (something that has to be done already).

I know nothing about CI but would it perform the compilation of the entire CMSSW code just to test a few additional custom packages?

**Comment by arizzi** [comment]
I think the idea was to move the VHbb part out of CMSSW so the test infrastructure should just prepare a release working area but not actually compile a full cmssw

**Comment by ferencek** [comment]
@arizzi, taking any analysis packages out of CMSSW and keeping them in a separate repo is in my experience more practical simply because you are not dragging along all sorts of unrelated developments in CMSSW itself. Also, switching to a more recent release in the same release cycle is a breeze without having to perform any rebasing and all other complications that come with it.

However, for CI I was expecting stuff to be done on the server side. On the other hand, if tests are done in a local area, then I don't see how using Gitlab is better compared to just sticking with GitHub.

**Comment by arizzi** [comment]
the question is if on the server side the "tests to run" are arbitrary code that can access scram repositories (being at CERN I guess so?)

**Comment by ferencek** [comment]
That I don't know. @jpata maybe knows more about what can be done as part of the tests infrastructure on the server side.

**Comment by arizzi** [comment]
clicking on his top thread link and then on the "compile" test
https://gitlab.cern.ch/ttH/CommonClassifier/builds/130564
it seems scram is there

**Comment by jpata** [comment]
yes, the single advantage and killer feature of CERN gitlab is that the CI runs directly on the CERN machines that have `cvmfs` and other standard things. Setting CMSSW up is really trivial, just look at this file: https://gitlab.cern.ch/ttH/CommonClassifier/blob/master/.gitlab-ci.yml

This doesn't compare with manual CI management via external services as on github.

**Comment by ferencek** [comment]
Looks nice :)

**Comment by jpata** [comment]
I have a build ongoing for my ttH code that also checks out vhbb: https://gitlab.cern.ch/jpata/tthbb13/pipelines/54522
Of course it takes a bit of time to compile our packages on a single core but one can do other things meanwhile. 
It doesn't compile CMSSW from scratch, released packages are taken from cvmfs.
